src/main/java/dev/jbang/cli/BaseBuildCommand.java

```diff
@@ -117,7 +122,7 @@ void build(Script script) throws IOException {
 		// always build the jar for native mode
 		// it allows integrations the options to produce the native image
 		if (!outjar.exists() || JavaUtil.javaVersion(requestedJavaVersion) < script.getBuildJdk()
-				|| nativeBuildRequired) {
```

Comment:
if you do:

`jbang build xyz.java` or `jbang xyz.java` and then do `jbang --fresh xyz.java` you will end up with the same cached jar folder; thus to be fresh that cached jar folder needs to be deleted before build happens.

Reply:
> if you do:
> 
> `jbang build xyz.java` or `jbang xyz.java` and then do `jbang --fresh xyz.java` you will end up with the same cached jar folder; thus to be fresh that cached jar folder needs to be deleted before build happens.

Did you test that? I tried and it worked.
It ends up in the same folder, but the files are replaced.
Here's the structure before and after `-f`
```
jars$ tree
.
├── xyz.java.0a7a631f89dd5ab9e831a8c7ccb9fa4ab103d2462cf92a5c2ea4ef39bc4fa7d4
│   ├── a.class
│   ├── META-INF
│   │   └── maven
│   │       └── g
│   │           └── a
│   │               └── v
│   │                   └── pom.xml
│   └── xyz.class
└── xyz.java.0a7a631f89dd5ab9e831a8c7ccb9fa4ab103d2462cf92a5c2ea4ef39bc4fa7d4.jar

6 directories, 4 files
leandro:jars$ tree
.
├── xyz.java.0a7a631f89dd5ab9e831a8c7ccb9fa4ab103d2462cf92a5c2ea4ef39bc4fa7d4
│   ├── a.class
│   ├── META-INF
│   │   └── maven
│   │       └── g
│   │           └── a
│   │               └── v
│   │                   └── pom.xml
│   └── xyz.class
└── xyz.java.0a7a631f89dd5ab9e831a8c7ccb9fa4ab103d2462cf92a5c2ea4ef39bc4fa7d4.jar

6 directories, 4 files
```

Output:
```
$ ./jbang/bin/jbang --verbose xyz.java
[jbang] System Java version detected as 11
[jbang] System Java version matches requested version 11
[jbang] Building jar...
[jbang] compile: /usr/lib/jvm/java-11-openjdk-amd64/bin/javac -d /home/leandro/.jbang/cache/jars/xyz.java.0a7a631f89dd5ab9e831a8c7ccb9fa4ab103d2462cf92a5c2ea4ef39bc4fa7d4 xyz.java /home/leandro/dev/githubprojects/tmp/jbang/build/distributions/a.java
[jbang] System Java version matches requested version 11
[jbang] run: /usr/lib/jvm/java-11-openjdk-amd64/bin/java -classpath /home/leandro/.jbang/cache/jars/xyz.java.0a7a631f89dd5ab9e831a8c7ccb9fa4ab103d2462cf92a5c2ea4ef39bc4fa7d4.jar: xyz
xyz!
a!
leandro:distributions$ vim a.java 
leandro:distributions$ ./jbang/bin/jbang -f --verbose xyz.java
[jbang] System Java version detected as 11
[jbang] System Java version matches requested version 11
[jbang] Building jar...
[jbang] compile: /usr/lib/jvm/java-11-openjdk-amd64/bin/javac -d /home/leandro/.jbang/cache/jars/xyz.java.0a7a631f89dd5ab9e831a8c7ccb9fa4ab103d2462cf92a5c2ea4ef39bc4fa7d4 xyz.java /home/leandro/dev/githubprojects/tmp/jbang/build/distributions/a.java
[jbang] System Java version matches requested version 11
[jbang] run: /usr/lib/jvm/java-11-openjdk-amd64/bin/java -classpath /home/leandro/.jbang/cache/jars/xyz.java.0a7a631f89dd5ab9e831a8c7ccb9fa4ab103d2462cf92a5c2ea4ef39bc4fa7d4.jar: xyz
xyz!
a!!!!
```

Reply:
> if you do:
> 
> `jbang build xyz.java` or `jbang xyz.java` and then do `jbang --fresh xyz.java` you will end up with the same cached jar folder; thus to be fresh that cached jar folder needs to be deleted before build happens.

opss. I see.
It the main file change it goes to a different folder.
It works, so delete is more to keep the cache clean.
Maybe it can be done in a different thread?

Reply:
> if you do:
> 
> `jbang build xyz.java` or `jbang xyz.java` and then do `jbang --fresh xyz.java` you will end up with the same cached jar folder; thus to be fresh that cached jar folder needs to be deleted before build happens.

So, if the main file changes it goes to a *different* folder, eg:
```
$ tree
.
├── xyz.java.0a7a631f89dd5ab9e831a8c7ccb9fa4ab103d2462cf92a5c2ea4ef39bc4fa7d4
│   ├── a.class
│   ├── META-INF
│   │   └── maven
│   │       └── g
│   │           └── a
│   │               └── v
│   │                   └── pom.xml
│   └── xyz.class
├── xyz.java.0a7a631f89dd5ab9e831a8c7ccb9fa4ab103d2462cf92a5c2ea4ef39bc4fa7d4.jar
├── xyz.java.fd496fd4a0345243df6a0f34bf2ba7d68aba36af1e767185bb58b291bc46f3db
│   ├── a.class
│   ├── META-INF
│   │   └── maven
│   │       └── g
│   │           └── a
│   │               └── v
│   │                   └── pom.xml
│   └── xyz.class
└── xyz.java.fd496fd4a0345243df6a0f34bf2ba7d68aba36af1e767185bb58b291bc46f3db.jar

12 directories, 8 files
```
We could delete the old jar file/folders, but I don't think we should do that.
As the hash changed, we would need to search all files/folders in the `jars` folder for the prefix (eg xyz.java*) and delete them.

Reply:
Would it perhaps be an option that the `build()` method first makes sure that the folder that's going to be used to do the compilation is empty? So instead of having different threads or clean-up phases we could just wipe the folder clean first and that way be sure that we always get a clean compilation. Or did I misunderstand and that wasn't the actual problem @maxandersen ?

Reply:
> Would it perhaps be an option that the `build()` method first makes sure that the folder that's going to be used to do the compilation is empty? So instead of having different threads or clean-up phases we could just wipe the folder clean first and that way be sure that we always get a clean compilation. Or did I misunderstand and that wasn't the actual problem @maxandersen ?

There are 2 problems I see if no clean up is done:
  1. Lots of unused/old jars in the cache folder: if the main class changes, then it's a different hash, hence a different jar/folder.
  2. Unused classes in the new jar: if the main class does **not** change, then it's the same jar/folder, and classes that *used to be* referenced in //SOURCES would remain. eg: a -> b -> c, then b does not use c anymore, c would remain.

We can solve these problems deleting both the jar/folder.

Reply:
> 1. Lots of unused/old jars in the cache folder: if the main class changes, then it's a different hash, hence a different jar/folder.

Although I think it's not possible to solve this one (when hash changes) the way Jbang works today.
For example this jar:
`Main.java.de29d2e6984d99c2750d0224c9b76f57e043c3012fc9a039d20e20d1e3c5a904.jar`

We can't simply delete by prefix (Main.java*), because the user can have lots of scripts called Main.java in different folders.

Reply:
Personally I don't care _that_ much about cleaning up the cache for things we no longer use. That's what `jbang cache clear` is for IMO. I just want to make sure that the _folder_ where we will do our compilation is clean. But that we could "fix" by simply running `Util.deleteFolder(tmpJarDir, true)` at the proper place (somewhere before `// add additional files`) in the `build()` method.

Reply:
my issue was not about many folders being created - that is separate issue and something to handle separately (we basically need a garbage collector :) 

Here i'm saying that currently the cache key is calculated by the main file thus if the main file is the same but some of the included resources have changed/moved then you can end up including left over files.

Thus we should have a check and a possible clean that ensure there are 0 files in the target classes dir before building starts happening.